go/arrow/array/dictionary.go

```diff
@@ -0,0 +1,1305 @@
+	case arrow.UINT8:
+		ret = hashing.NewUint8MemoTable(0)
+	case arrow.INT16:
+		ret = hashing.NewInt16MemoTable(0)
+	case arrow.UINT16:
+		ret = hashing.NewUint16MemoTable(0)
+	case arrow.INT32:
+		ret = hashing.NewInt32MemoTable(0)
+	case arrow.UINT32:
+		ret = hashing.NewUint32MemoTable(0)
+	case arrow.INT64:
+		ret = hashing.NewInt64MemoTable(0)
+	case arrow.UINT64:
+		ret = hashing.NewUint64MemoTable(0)
+	case arrow.DURATION, arrow.TIMESTAMP, arrow.DATE64, arrow.TIME64:
+		ret = hashing.NewInt64MemoTable(0)
+	case arrow.TIME32, arrow.DATE32, arrow.INTERVAL_MONTHS:
+		ret = hashing.NewInt32MemoTable(0)
+	case arrow.FLOAT16:
+		ret = hashing.NewUint16MemoTable(0)
+	case arrow.FLOAT32:
+		ret = hashing.NewFloat32MemoTable(0)
+	case arrow.FLOAT64:
+		ret = hashing.NewFloat64MemoTable(0)
+	case arrow.BINARY, arrow.FIXED_SIZE_BINARY, arrow.DECIMAL128, arrow.INTERVAL_DAY_TIME, arrow.INTERVAL_MONTH_DAY_NANO:
+		ret = hashing.NewBinaryMemoTable(0, 0, NewBinaryBuilder(mem, arrow.BinaryTypes.Binary))
+	case arrow.STRING:
+		ret = hashing.NewBinaryMemoTable(0, 0, NewBinaryBuilder(mem, arrow.BinaryTypes.String))
+	case arrow.NULL:
+	default:
+		debug.Assert(false, "unimplemented dictionary value type")
+		err = fmt.Errorf("unimplemented dictionary value type, %s", dt)
+	}
+
+	return
+}
+
+type DictionaryBuilder interface {
+	Builder
+
+	NewDictionaryArray() *Dictionary
+	NewDelta() (indices, delta Interface, err error)
+	AppendArray(Interface) error
+	ResetFull()
+}
+
+type dictionaryBuilder struct {
+	builder
+
+	dt          *arrow.DictionaryType
+	deltaOffset int
+	memoTable   hashing.MemoTable
+	idxBuilder  indexBuilder
+}
+
+func NewDictionaryBuilderWithDict(mem memory.Allocator, dt *arrow.DictionaryType, init Interface) DictionaryBuilder {
+	if init != nil && !arrow.TypeEqual(dt.ValueType, init.DataType()) {
+		panic(fmt.Errorf("arrow/array: cannot initialize dictionary type %T with array of type %T", dt.ValueType, init.DataType()))
+	}
+
+	idxbldr, err := createIndexBuilder(mem, dt.IndexType.(arrow.FixedWidthDataType))
+	if err != nil {
+		panic(fmt.Errorf("arrow/array: unsupported builder for index type of %T", dt))
+	}
+
+	memo, err := createMemoTable(mem, dt.ValueType)
+	if err != nil {
+		panic(fmt.Errorf("arrow/array: unsupported builder for value type of %T", dt))
+	}
+
+	bldr := dictionaryBuilder{
+		builder:    builder{refCount: 1, mem: mem},
+		idxBuilder: idxbldr,
+		memoTable:  memo,
+		dt:         dt,
+	}
+
+	switch dt.ValueType.ID() {
+	case arrow.NULL:
+		ret := &NullDictionaryBuilder{bldr}
+		debug.Assert(init == nil, "arrow/array: doesn't make sense to init a null dictionary")
+		return ret
+	case arrow.UINT8:
+		ret := &Uint8DictionaryBuilder{bldr}
+		if init != nil {
+			if err = ret.InsertDictValues(init.(*Uint8)); err != nil {
+				panic(err)
+			}
+		}
+		return ret
+	case arrow.INT8:
+		ret := &Int8DictionaryBuilder{bldr}
+		if init != nil {
+			if err = ret.InsertDictValues(init.(*Int8)); err != nil {
+				panic(err)
```

Comment:
why is panic the right thing here?

Reply:
The alternative would be to change the function signature to return `(DictionaryBuilder, error)` but all of the other functions to construct a NewBuilder currently would panic if they ran into an error and don't return `(Builder, error)` so I was following the existing pattern.

Reply:
would it make sense to clean this up at some point later?

Reply:
probably. 

Honestly, I'd like to replace a lot of this using Go generics that came out with go1.18, but since we want to maintain support for at least `CurrentVersion`-1 I'm going to have to wait until Go1.19 is released before doing so (because switching to Go generics would make this incompatible with any Go versions <1.18).